HTTP chain observer 0.19

ch1bo opened the pull request:
Backport of the HTTP chain observation to build a docker image of `hydra-chain-observer` for version `0.19` and deploy it to the explorer.

This could be released as `0.19.1`

---

* [ ] CHANGELOG updated or not needed
* [ ] Documentation updated or not needed
* [ ] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained herafter

ch1bo committed:
Draft desired behavior to report observations

ch1bo committed:
Use options instead of subcommands in hydra-chain-observer

This makes documentation and implementation for common options like
"--start-chain-from" and (new) "--explorer" less redundant.

ch1bo committed:
Add --explorer to hydra-chain-observer

ch1bo committed:
Perform POST requests on observation

ch1bo committed:
Change ChainObservation to match observer API json schema

ch1bo committed:
Add FromJSON OnChainTx

This type is used by hydra-chain-observer and hydra-explorer
respectively on the observations API

ch1bo committed:
Add ToJSON/FromJSON orphan instances for NetworkMagic

Cardano-api has instances for Network, but not NetworkMagic.

ch1bo committed:
Add Arbitrary NetworkId instance to hydra-cardano-api

ch1bo committed:
Add roundtrip golden tests for OnChainTx

This data type is essential for the hydra-chain-observer ->
hydra-explorer API

ch1bo committed:
Submit version and network to hydra-explorer

The hydra-chain-observer now uses the hydraNodeVersion, which is either
patched into it or uses the git revision from the repository, and
reports this along with a network id to the explorer.

ch1bo committed:
Update changelog

Update changelog and hydra-chain-observer version

Using patch version component as this is not really changing the main
hydra features, but only the reporting into the hydra-explorer.

ch1bo committed:
Fix hydra-chain-observer network path param

The show instance of NetworkMagic is not just the number.

ch1bo committed:
Add headId to singlePartyOpenAHead scenario callback

This allows us to easily open heads and get their head id.

ch1bo committed:
TMP: Build docker image of hydra-chain-observer

ch1bo committed:
Add docker image for hydra-chain-observer

This will be available as a flake output and built by CI into https://github.com/orgs/cardano-scaling/packages

ch1bo committed:
Remove legacy hydra-explorer from repo

noonio reviewed (approved):
Happy to approve despite the failing Mithril tests.

ch1bo commented:
> Happy to approve despite the failing Mithril tests.

Thanks, but I'm not yet fully convinced this actually works :o .. still have not seen a `0.19.0` head on http://explorer.hydra.family/ :thinking:

ch1bo committed:
Fix decoding of Hydra keys and ignore problematic head init

We were not correctly decoding Hydra verification keys (but crashing via
error) because of a missing data constructor.

This also makes the observeInitTx sort out heads that would have such
problematic party datums.

An example transaction with problematic data is:
b860a236a7e77577628bf705286d449188831cf90d714934f1cc06369c6e3953 on preprod

ch1bo commented on hydra-chain-observer/src/Hydra/ChainObserver.hs:
```diff
@@ -51,37 +53,55 @@ import Ouroboros.Network.Protocol.ChainSync.Client (
+data ChainObservation = ChainObservation
+  { point :: ChainPoint
+  , blockNo :: BlockNo
+  , onChainTx :: Maybe (OnChainTx Tx)
```
@noonio Instead of https://github.com/cardano-scaling/hydra-explorer/pull/32 we should do just

```suggestion
  , observedTx :: Maybe (OnChainTx Tx)
```

ch1bo commented on hydra-chain-observer/golden/ChainObservation.json:
```diff
@@ -0,0 +1,922 @@
+    "samples": [
+        {
+            "blockNo": 7783348,
+            "observedTx": {
```
@noonio Surprisingly this is correct. So where the golden tests not failing?

ch1bo commented on hydra-chain-observer/test/Hydra/ChainObserverSpec.hs:
```diff
@@ -3,18 +3,31 @@ module Hydra.ChainObserverSpec where
   parallel $ do
+    -- NOTE: Detect regressions in interface to hydra-explorer
+    let settings = defaultSettings{sampleSize = 1}
+    roundtripAndGoldenSpecsWithSettings settings $ Proxy @ChainObservation
```
@noonio This golden test in particular?

ch1bo replied:
:facepalm: The `hydra-chain-observer` test suite is not run in CI

noonio replied:
🥲

ch1bo reviewed (commented):
Comments to fix what @noonio found in https://github.com/cardano-scaling/hydra-explorer/pull/32

noonio committed:
Update hydra-chain-observer/src/Hydra/ChainObserver.hs

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

noonio commented:
I think this is ready to merge now; thoughts @ch1bo ?
